src/Interpreters/ActionsDAG.cpp

```diff
@@ -628,13 +628,17 @@ static ColumnWithTypeAndName executeActionForHeader(const ActionsDAG::Node * nod
             if (!array)
                 throw Exception(ErrorCodes::TYPE_MISMATCH,
                                 "ARRAY JOIN of not array nor map: {}", node->result_name);
-            res_column.column = array->getDataPtr()->cloneEmpty();
+            res_column.column = array->getDataPtr();
+            if (input_rows_count < array->size())
+                res_column.column = res_column.column->cloneResized(array->getOffsets()[input_rows_count - 1]);
```

Comment:
This line is hard to understand. Why do we resite to the value of the offset?

Reply:
This is how `arrayJoin()` works - it takes a `ColumnArray` and returns its data. So if only `input_rows_count` are to be used, then we should take only the first part of the data.

So this line is the same as `array->cloneResized(input_rows_count)->getDataPtr())` but in more optimized way. I'll add a comment about that here.

Reply:
Is it possible `input_rows_count` is zero here?

Reply:
I've moved this very `ColumnArray`-specific code to the `ColumnArray` class itself, now it looks better.